File: 2751-robot-collisions/2751-robot-collisions.cpp
class Solution {
public:
    bool compare(int x, int y, vector<int>& pos) {
        return pos[x] < pos[y];
    }

    vector<int> survivedRobotsHealths(vector<int>& pos, vector<int>& healths, string dir) {
        stack <int> st;
        int num_of_robots = pos.size();
        vector <int> robots(num_of_robots);

        for (int i = 0; i < num_of_robots; i++) {
            robots[i] = i;
        }

        sort(robots.begin(), robots.end(), [&](int a, int b) {
            return compare(a, b, pos);
        });

        for (int idx : robots) {
            if (dir[idx] == 'R') {
                st.push(idx);
            } else {
                while (!st.empty() && healths[idx] > 0) {
                    int right = st.top();
                    if (healths[idx] < healths[right]) {
                        healths[idx] = 0;
                        healths[right] -= 1;
                    } else if (healths[idx] > healths[right]) {
                        healths[idx] -= 1;
                        healths[right] = 0;
                        st.pop();
                    } else {
                        healths[idx] = 0;
                        healths[right] = 0;
                        st.pop();
                    }
                }
            }
        }

        vector <int> ans;
        for (int health : healths) {
            if (health > 0) {
                ans.push_back(health);
            }
        }

        return ans;
    }
};

File: 2751-robot-collisions/2751-robot-collisions.py
class Solution:
    def survivedRobotsHealths(self, pos: List[int], healths: List[int], directions: str) -> List[int]:
        stack = []
        num_of_robots = len(pos)
        robots = [i for i in range(num_of_robots)]

        robots.sort(key = lambda x: pos[x])

        for i in robots:
            if directions[i] == 'R':
                stack.append(i)
            else:
                while stack and healths[i] > 0:
                    top = stack[-1]
                    if healths[i] > healths[top]:
                        healths[i] -= 1
                        healths[top] = 0
                        stack.pop()
                    elif healths[i] < healths[top]:
                        healths[i] = 0
                        healths[top] -= 1
                    else:
                        healths[i] = 0
                        healths[top] = 0
                        stack.pop()

        return [health for health in healths if health]